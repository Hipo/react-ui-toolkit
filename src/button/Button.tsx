import React from "react";
import classNames from "classnames";

import Spinner from "../spinner/Spinner";

// SCSS import is moved here to be able to override spinner styles without nesting
import "./_button.scss";

export interface ButtonProps {
  testid?: string;
  children: React.ReactNode;
  onClick?: React.ReactEventHandler<HTMLButtonElement>;
  onMouseOver?: React.ReactEventHandler<HTMLButtonElement>;
  onMouseDown?: React.ReactEventHandler<HTMLButtonElement>;
  onMouseUp?: React.ReactEventHandler<HTMLButtonElement>;
  onFocus?: React.ReactEventHandler<HTMLButtonElement>;
  onBlur?: React.ReactEventHandler<HTMLButtonElement>;
  id?: string;
  lang?: string;
  type?: "button" | "submit" | "reset";
  shouldDisplaySpinner?: boolean;
  isDisabled?: boolean;
  shouldPreventDefault?: boolean;
  shouldStopPropagation?: boolean;
  shouldFocus?: boolean;
  ariaLabel?: string;
  customClassName?: string;
  tabIndex?: number;
}

function Button({
  testid,
  type = "button",
  id,
  lang,
  onClick,
  children,
  customClassName,
  shouldPreventDefault = true,
  shouldStopPropagation = true,
  shouldFocus,
  shouldDisplaySpinner,
  ariaLabel,
  isDisabled,
  onMouseOver,
  onMouseDown,
  onMouseUp,
  onFocus,
  onBlur,
  tabIndex
}: ButtonProps) {
  const isButtonDisabled = Boolean(isDisabled || shouldDisplaySpinner);
  const containerClassName = classNames("button", customClassName, {
    "button--is-inactive": isButtonDisabled
  });

  return (
    <button
      id={id}
      data-testid={testid}
      tabIndex={tabIndex}
      className={containerClassName}
      type={type}
      lang={lang}
      // eslint-disable-next-line jsx-a11y/no-autofocus
      autoFocus={shouldFocus}
      onClick={handleClick}
      onMouseOver={handleMouseOver}
      onMouseDown={onMouseDown}
      onMouseUp={onMouseUp}
      onFocus={handleFocus}
      onBlur={onBlur}
      disabled={isButtonDisabled}
      aria-label={ariaLabel}>
      {children}

      {shouldDisplaySpinner && (
        <div className={"button__spinner-container"}>
          <Spinner
            customClassName={"button__spinner"}
            aria-label={"Button spinner visible. Button inactivated."}
          />
        </div>
      )}
    </button>
  );

  function handleClick(event: React.SyntheticEvent<HTMLButtonElement>) {
    if (onClick) {
      if (shouldPreventDefault) {
        event.preventDefault();
      }

      if (shouldStopPropagation) {
        event.stopPropagation();
      }

      if (!isButtonDisabled) {
        onClick(event);
      }
    }
  }

  function handleMouseOver(event: React.SyntheticEvent<HTMLButtonElement>) {
    if (onMouseOver) {
      onMouseOver(event);
    }
  }

  function handleFocus(event: React.SyntheticEvent<HTMLButtonElement>) {
    if (onFocus) {
      onFocus(event);
    }
  }
}

export default Button;
