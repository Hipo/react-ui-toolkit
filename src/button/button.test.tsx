import React from "react";
import {render, screen, fireEvent} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";

import Button, {ButtonProps} from "./Button";
import {testA11y} from "../core/utils/test/testUtils";

describe("<Button />", () => {
  const defaultButtonProps: ButtonProps = {
    testid: "button",
    children: "Test"
  };

  it("should render correctly", () => {
    render(<Button {...defaultButtonProps} />);
  });

  it("should render children correctly", () => {
    render(<Button {...defaultButtonProps} />);

    expect(screen.getByRole("button")).toHaveTextContent("Test");
  });

  it("should pass a11y test", async () => {
    const {container} = render(<Button {...defaultButtonProps} />);

    await testA11y(container);
  });

  it("should add disabled attribute when isDisabled is true", () => {
    render(<Button isDisabled={true} {...defaultButtonProps} />);

    expect(screen.getByRole("button")).toBeDisabled();
  });

  it("should display custom spinner correctly", () => {
    const customSpinnerContent = <p data-testid={"custom-spinner"}>{"Loading..."}</p>;

    const {rerender, container} = render(
      <Button
        shouldDisplaySpinner={true}
        customSpinner={customSpinnerContent}
        {...defaultButtonProps}
      />
    );

    const customSpinner = screen.getByText("Loading...");

    expect(container).toContainElement(customSpinner);

    rerender(
      <Button
        shouldDisplaySpinner={false}
        customSpinner={customSpinnerContent}
        {...defaultButtonProps}
      />
    );

    expect(container).not.toContainElement(customSpinner);
  });

  it("should not run click event handler while button is disabled", () => {
    const handleClick = jest.fn();

    render(<Button onClick={handleClick} isDisabled={true} {...defaultButtonProps} />);

    userEvent.click(screen.getByRole("button"));
    expect(handleClick).not.toHaveBeenCalled();
  });

  it("should not run click event handler while shouldDisplaySpinner is true", () => {
    const handleClick = jest.fn();

    render(
      <Button onClick={handleClick} shouldDisplaySpinner={true} {...defaultButtonProps} />
    );

    userEvent.click(screen.getByRole("button"));
    expect(handleClick).not.toHaveBeenCalled();
  });

  it("should run click event handler correctly", () => {
    const handleClick = jest.fn();

    render(<Button onClick={handleClick} {...defaultButtonProps} />);

    userEvent.click(screen.getByRole("button"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it("should run mouseOver event handler correctly", () => {
    const handleMouseOver = jest.fn();
    const handleFocus = jest.fn();

    render(
      <Button
        onFocus={handleFocus}
        onMouseOver={handleMouseOver}
        {...defaultButtonProps}
      />
    );

    fireEvent.mouseOver(screen.getByRole("button"));
    expect(handleMouseOver).toHaveBeenCalledTimes(1);
  });

  it("should run focus event handler correctly", () => {
    const handleFocus = jest.fn();

    render(<Button onFocus={handleFocus} {...defaultButtonProps} />);

    fireEvent.focus(screen.getByRole("button"));
    expect(handleFocus).toHaveBeenCalledTimes(1);
  });

  it("should run mouseUp event handler correctly", () => {
    const handleMouseUp = jest.fn();

    render(<Button onMouseUp={handleMouseUp} {...defaultButtonProps} />);

    fireEvent.mouseUp(screen.getByRole("button"));
    expect(handleMouseUp).toHaveBeenCalledTimes(1);
  });

  it("should run mouseDown event handler correctly", () => {
    const handleMouseDown = jest.fn();

    render(<Button onMouseDown={handleMouseDown} {...defaultButtonProps} />);

    fireEvent.mouseDown(screen.getByRole("button"));
    expect(handleMouseDown).toHaveBeenCalledTimes(1);
  });

  it("should run blur event handler correctly", () => {
    const handleBlur = jest.fn();

    render(<Button onBlur={handleBlur} {...defaultButtonProps} />);

    fireEvent.blur(screen.getByRole("button"));
    expect(handleBlur).toHaveBeenCalledTimes(1);
  });

  it("should has the proper type attribute", () => {
    const {rerender} = render(<Button {...defaultButtonProps} />);

    expect(screen.getByRole("button")).toHaveAttribute("type", "button");

    rerender(<Button type={"submit"} {...defaultButtonProps} />);

    expect(screen.getByRole("button")).toHaveAttribute("type", "submit");

    rerender(<Button type={"reset"} {...defaultButtonProps} />);

    expect(screen.getByRole("button")).toHaveAttribute("type", "reset");
  });
});
