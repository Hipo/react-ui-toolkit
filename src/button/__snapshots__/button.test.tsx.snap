// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should match snapshot 1`] = `
<button
  className="button"
  data-testid="button"
  disabled={false}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseOver={[Function]}
  type="button"
>
  Test
</button>
`;
